File: frontend/src/types/watchlist.ts
import { z } from "zod";

export const WatchlistMetadataSchema = z.object({
  news_count: z.number().nullish(),
  earnings_date: z.string().nullish(),
  iv30_percentile: z.number().nullish(),
  options_volume_ratio: z.number().nullish(),
  sentiment_score: z.number().nullish(),
});

export const WatchlistItemSchema = z.object({
  symbol: z.string(),
  price: z.number(),
  change_pct: z.number(),
  subtitle: z.string(),
  metadata: WatchlistMetadataSchema.default({}),
});

export const WatchlistResponseSchema = z.object({
  hot_now: z.array(WatchlistItemSchema),
  earnings: z.array(WatchlistItemSchema),
  unusual_options: z.array(WatchlistItemSchema),
  sentiment_up: z.array(WatchlistItemSchema),
  sentiment_down: z.array(WatchlistItemSchema),
  notes: z.record(z.string()).default({}),
  updated_at: z.string(),
});

export type WatchlistItem = z.infer<typeof WatchlistItemSchema>;
export type WatchlistResponse = z.infer<typeof WatchlistResponseSchema>;
export type WatchlistCategoryKey =
  | "hot_now"
  | "earnings"
  | "unusual_options"
  | "sentiment_up"
  | "sentiment_down";

File: backend/schemas/watchlist.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class WatchlistMetadata(BaseModel):
    """Per-category extras. All optional — populated only when relevant."""

    model_config = ConfigDict(extra="ignore")

    news_count: int | None = None
    earnings_date: str | None = None
    iv30_percentile: float | None = None
    options_volume_ratio: float | None = None
    sentiment_score: float | None = None


class WatchlistItemOut(BaseModel):
    symbol: str
    price: float
    change_pct: float
    subtitle: str
    metadata: WatchlistMetadata = Field(default_factory=WatchlistMetadata)


class WatchlistResponse(BaseModel):
    hot_now: list[WatchlistItemOut]
    earnings: list[WatchlistItemOut]
    unusual_options: list[WatchlistItemOut]
    sentiment_up: list[WatchlistItemOut]
    sentiment_down: list[WatchlistItemOut]
    # Optional per-category placeholder text. Surfaces things like
    # "Sentiment unavailable on free tier" instead of the generic empty state.
    notes: dict[str, str] = Field(default_factory=dict)
    updated_at: datetime
